File: examples/Algorithm/python/13_topological_sort.py
"""
Topological Sort
Topological Sorting

Arranges vertices of a Directed Acyclic Graph (DAG) in linear order.
"""

from typing import List, Dict, Set, Optional, Tuple
from collections import deque, defaultdict


# =============================================================================
# 1. Kahn's Algorithm (BFS-based)
# =============================================================================

def topological_sort_kahn(n: int, edges: List[Tuple[int, int]]) -> List[int]:
    """
    Kahn's Algorithm (in-degree based)
    Time Complexity: O(V + E)
    Space Complexity: O(V + E)

    edges: [(from, to), ...] - from -> to dependency
    Returns: topologically sorted order, empty list if cycle exists
    """
    # Build graph and in-degree array
    graph = defaultdict(list)
    in_degree = [0] * n

    for u, v in edges:
        graph[u].append(v)
        in_degree[v] += 1

    # Start with nodes having in-degree 0
    queue = deque([i for i in range(n) if in_degree[i] == 0])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)

        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    # Check for cycle by verifying all nodes were visited
    return result if len(result) == n else []


# =============================================================================
# 2. DFS-based Topological Sort
# =============================================================================

def topological_sort_dfs(n: int, edges: List[Tuple[int, int]]) -> List[int]:
    """
    DFS-based Topological Sort
    Time Complexity: O(V + E)
    Space Complexity: O(V + E)
    """
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    WHITE, GRAY, BLACK = 0, 1, 2
    color = [WHITE] * n
    result = []
    has_cycle = False

    def dfs(node: int) -> None:
        nonlocal has_cycle

        if has_cycle:
            return

        color[node] = GRAY  # In progress

        for neighbor in graph[node]:
            if color[neighbor] == GRAY:  # Cycle detected
                has_cycle = True
                return
            if color[neighbor] == WHITE:
                dfs(neighbor)

        color[node] = BLACK  # Completed
        result.append(node)

    for i in range(n):
        if color[i] == WHITE:
            dfs(i)

    return result[::-1] if not has_cycle else []


# =============================================================================
# 3. Cycle Detection
# =============================================================================

def has_cycle(n: int, edges: List[Tuple[int, int]]) -> bool:
    """
    Check for cycle in a directed graph
    Time Complexity: O(V + E)
    """
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    WHITE, GRAY, BLACK = 0, 1, 2
    color = [WHITE] * n

    def dfs(node: int) -> bool:
        color[node] = GRAY

        for neighbor in graph[node]:
            if color[neighbor] == GRAY:
                return True  # Cycle
            if color[neighbor] == WHITE and dfs(neighbor):
                return True

        color[node] = BLACK
        return False

    for i in range(n):
        if color[i] == WHITE and dfs(i):
            return True

    return False


# =============================================================================
# 4. Find All Topological Sort Orders
# =============================================================================

def all_topological_sorts(n: int, edges: List[Tuple[int, int]]) -> List[List[int]]:
    """
    Find all possible topological sort orders
    Time Complexity: O(V! * (V + E)) - worst case
    """
    graph = defaultdict(list)
    in_degree = [0] * n

    for u, v in edges:
        graph[u].append(v)
        in_degree[v] += 1

    result = []
    path = []
    visited = [False] * n

    def backtrack():
        # All nodes visited
        if len(path) == n:
            result.append(path.copy())
            return

        for i in range(n):
            if not visited[i] and in_degree[i] == 0:
                # Choose
                visited[i] = True
                path.append(i)
                for neighbor in graph[i]:
                    in_degree[neighbor] -= 1

                backtrack()

                # Restore
                visited[i] = False
                path.pop()
                for neighbor in graph[i]:
                    in_degree[neighbor] += 1

    backtrack()
    return result


# =============================================================================
# 5. Practical Problem: Course Schedule
# =============================================================================

def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    """
    Check if all courses can be completed (no cycle = possible)
    prerequisites: [course, prereq] - prereq -> course
    """
    edges = [(prereq, course) for course, prereq in prerequisites]
    return not has_cycle(num_courses, edges)


def find_order(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    """
    Return the course completion order
    """
    edges = [(prereq, course) for course, prereq in prerequisites]
    return topological_sort_kahn(num_courses, edges)


# =============================================================================
# 6. Practical Problem: Alien Dictionary
# =============================================================================

def alien_order(words: List[str]) -> str:
    """
    Determine alien alphabet order
    Assumes the word list is sorted in lexicographic order
    """
    # Collect all characters
    chars = set()
    for word in words:
        chars.update(word)

    # Build graph
    graph = defaultdict(set)
    in_degree = defaultdict(int)
    for char in chars:
        in_degree[char] = 0

    for i in range(len(words) - 1):
        word1, word2 = words[i], words[i + 1]

        # Check for invalid ordering
        if len(word1) > len(word2) and word1.startswith(word2):
            return ""

        # Determine order from first different character
        for c1, c2 in zip(word1, word2):
            if c1 != c2:
                if c2 not in graph[c1]:
                    graph[c1].add(c2)
                    in_degree[c2] += 1
                break

    # Topological sort
    queue = deque([c for c in chars if in_degree[c] == 0])
    result = []

    while queue:
        char = queue.popleft()
        result.append(char)

        for neighbor in graph[char]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return ''.join(result) if len(result) == len(chars) else ""


# =============================================================================
# 7. Practical Problem: Parallel Courses
# =============================================================================

def minimum_semesters(n: int, relations: List[List[int]]) -> int:
    """
    Minimum number of semesters to complete all courses
    Parallel enrollment allowed, relations: [prev, next]
    """
    graph = defaultdict(list)
    in_degree = [0] * (n + 1)

    for prev_course, next_course in relations:
        graph[prev_course].append(next_course)
        in_degree[next_course] += 1

    # Start with nodes having in-degree 0 (1-indexed)
    queue = deque([i for i in range(1, n + 1) if in_degree[i] == 0])
    semesters = 0
    completed = 0

    while queue:
        semesters += 1
        next_queue = deque()

        while queue:
            course = queue.popleft()
            completed += 1

            for next_course in graph[course]:
                in_degree[next_course] -= 1
                if in_degree[next_course] == 0:
                    next_queue.append(next_course)

        queue = next_queue

    return semesters if completed == n else -1


# =============================================================================
# 8. Practical Problem: Longest Path in DAG
# =============================================================================

def longest_path_dag(n: int, edges: List[Tuple[int, int, int]]) -> List[int]:
    """
    Longest path to each node in a DAG (with weights)
    edges: [(from, to, weight), ...]
    """
    graph = defaultdict(list)
    in_degree = [0] * n

    for u, v, w in edges:
        graph[u].append((v, w))
        in_degree[v] += 1

    # Topological sort
    topo_order = []
    queue = deque([i for i in range(n) if in_degree[i] == 0])

    while queue:
        node = queue.popleft()
        topo_order.append(node)
        for neighbor, _ in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    # Calculate longest path
    dist = [0] * n

    for node in topo_order:
        for neighbor, weight in graph[node]:
            dist[neighbor] = max(dist[neighbor], dist[node] + weight)

    return dist


# =============================================================================
# 9. Practical Problem: Build Order
# =============================================================================

def build_order(projects: List[str], dependencies: List[Tuple[str, str]]) -> List[str]:
    """
    Determine build order
    dependencies: [(proj, depends_on), ...] - proj depends on depends_on
    """
    # Project index mapping
    proj_to_idx = {p: i for i, p in enumerate(projects)}
    n = len(projects)

    # Convert edges (depends_on -> proj)
    edges = [(proj_to_idx[dep], proj_to_idx[proj]) for proj, dep in dependencies]

    # Topological sort
    order = topological_sort_kahn(n, edges)

    if not order:
        return []  # Cycle exists

    return [projects[i] for i in order]


# =============================================================================
# Tests
# =============================================================================

def main():
    print("=" * 60)
    print("Topological Sort Examples")
    print("=" * 60)

    # 1. Kahn's Algorithm
    print("\n[1] Kahn's Algorithm (BFS)")
    n = 6
    edges = [(5, 2), (5, 0), (4, 0), (4, 1), (2, 3), (3, 1)]
    result = topological_sort_kahn(n, edges)
    print(f"    Nodes: 0-5, Edges: {edges}")
    print(f"    Topological sort: {result}")

    # 2. DFS-based
    print("\n[2] DFS-based Topological Sort")
    result = topological_sort_dfs(n, edges)
    print(f"    Topological sort: {result}")

    # 3. Cycle Detection
    print("\n[3] Cycle Detection")
    cyclic_edges = [(0, 1), (1, 2), (2, 0)]
    acyclic_edges = [(0, 1), (1, 2), (0, 2)]
    print(f"    {cyclic_edges}: cycle {has_cycle(3, cyclic_edges)}")
    print(f"    {acyclic_edges}: cycle {has_cycle(3, acyclic_edges)}")

    # 4. All Topological Sorts
    print("\n[4] All Topological Sort Orders")
    n = 4
    edges = [(0, 1), (0, 2), (1, 3), (2, 3)]
    all_orders = all_topological_sorts(n, edges)
    print(f"    Nodes: 0-3, Edges: {edges}")
    print(f"    All orders: {all_orders}")

    # 5. Course Schedule
    print("\n[5] Course Schedule")
    num_courses = 4
    prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]]
    can = can_finish(num_courses, prerequisites)
    order = find_order(num_courses, prerequisites)
    print(f"    Courses: {num_courses}, Prerequisites: {prerequisites}")
    print(f"    Can finish: {can}")
    print(f"    Order: {order}")

    # 6. Alien Dictionary
    print("\n[6] Alien Dictionary")
    words = ["wrt", "wrf", "er", "ett", "rftt"]
    order = alien_order(words)
    print(f"    Words: {words}")
    print(f"    Alphabet order: {order}")

    # 7. Minimum Semesters
    print("\n[7] Minimum Semesters")
    n = 3
    relations = [[1, 3], [2, 3]]
    semesters = minimum_semesters(n, relations)
    print(f"    Courses: {n}, Relations: {relations}")
    print(f"    Minimum semesters: {semesters}")

    # 8. Longest Path in DAG
    print("\n[8] Longest Path in DAG")
    n = 4
    edges = [(0, 1, 3), (0, 2, 2), (1, 3, 4), (2, 3, 1)]
    dist = longest_path_dag(n, edges)
    print(f"    Edges: {edges}")
    print(f"    Longest distance to each node: {dist}")

    # 9. Build Order
    print("\n[9] Build Order")
    projects = ['a', 'b', 'c', 'd', 'e', 'f']
    dependencies = [('d', 'a'), ('b', 'f'), ('d', 'b'), ('a', 'f'), ('c', 'd')]
    order = build_order(projects, dependencies)
    print(f"    Projects: {projects}")
    print(f"    Dependencies: {dependencies}")
    print(f"    Build order: {order}")

    print("\n" + "=" * 60)


if __name__ == "__main__":
    main()

File: examples/Algorithm/c/13_topological_sort.c
/*
 * Topological Sort
 * Kahn's Algorithm, DFS-based, Cycle Detection
 *
 * Finds a linear ordering of vertices in a Directed Acyclic Graph (DAG).
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/* =============================================================================
 * 1. Graph Structure
 * ============================================================================= */

typedef struct AdjNode {
    int vertex;
    struct AdjNode* next;
} AdjNode;

typedef struct {
    AdjNode** adj;
    int vertices;
} Graph;

Graph* graph_create(int vertices) {
    Graph* g = malloc(sizeof(Graph));
    g->vertices = vertices;
    g->adj = calloc(vertices, sizeof(AdjNode*));
    return g;
}

void graph_add_edge(Graph* g, int src, int dest) {
    AdjNode* node = malloc(sizeof(AdjNode));
    node->vertex = dest;
    node->next = g->adj[src];
    g->adj[src] = node;
}

void graph_free(Graph* g) {
    for (int i = 0; i < g->vertices; i++) {
        AdjNode* node = g->adj[i];
        while (node) {
            AdjNode* temp = node;
            node = node->next;
            free(temp);
        }
    }
    free(g->adj);
    free(g);
}

/* =============================================================================
 * 2. Kahn's Algorithm (BFS-based)
 * ============================================================================= */

int* kahn_topological_sort(Graph* g, bool* has_cycle) {
    int* in_degree = calloc(g->vertices, sizeof(int));
    int* result = malloc(g->vertices * sizeof(int));
    int result_idx = 0;

    /* Calculate in-degrees */
    for (int v = 0; v < g->vertices; v++) {
        AdjNode* node = g->adj[v];
        while (node) {
            in_degree[node->vertex]++;
            node = node->next;
        }
    }

    /* Add vertices with in-degree 0 to queue */
    int* queue = malloc(g->vertices * sizeof(int));
    int front = 0, rear = 0;

    for (int i = 0; i < g->vertices; i++) {
        if (in_degree[i] == 0)
            queue[rear++] = i;
    }

    /* BFS */
    while (front < rear) {
        int v = queue[front++];
        result[result_idx++] = v;

        AdjNode* node = g->adj[v];
        while (node) {
            in_degree[node->vertex]--;
            if (in_degree[node->vertex] == 0)
                queue[rear++] = node->vertex;
            node = node->next;
        }
    }

    *has_cycle = (result_idx != g->vertices);

    free(in_degree);
    free(queue);
    return result;
}

/* =============================================================================
 * 3. DFS-based Topological Sort
 * ============================================================================= */

void dfs_topological_util(Graph* g, int v, bool visited[], int stack[], int* stack_top) {
    visited[v] = true;

    AdjNode* node = g->adj[v];
    while (node) {
        if (!visited[node->vertex])
            dfs_topological_util(g, node->vertex, visited, stack, stack_top);
        node = node->next;
    }

    stack[(*stack_top)++] = v;
}

int* dfs_topological_sort(Graph* g) {
    bool* visited = calloc(g->vertices, sizeof(bool));
    int* stack = malloc(g->vertices * sizeof(int));
    int stack_top = 0;

    for (int i = 0; i < g->vertices; i++) {
        if (!visited[i])
            dfs_topological_util(g, i, visited, stack, &stack_top);
    }

    /* Generate result in reverse stack order */
    int* result = malloc(g->vertices * sizeof(int));
    for (int i = 0; i < g->vertices; i++)
        result[i] = stack[g->vertices - 1 - i];

    free(visited);
    free(stack);
    return result;
}

/* =============================================================================
 * 4. DFS Topological Sort with Cycle Detection
 * ============================================================================= */

typedef enum { WHITE, GRAY, BLACK } Color;

bool dfs_topological_cycle_util(Graph* g, int v, Color color[], int stack[], int* stack_top) {
    color[v] = GRAY;

    AdjNode* node = g->adj[v];
    while (node) {
        if (color[node->vertex] == GRAY)
            return true;  /* Cycle detected */
        if (color[node->vertex] == WHITE) {
            if (dfs_topological_cycle_util(g, node->vertex, color, stack, stack_top))
                return true;
        }
        node = node->next;
    }

    color[v] = BLACK;
    stack[(*stack_top)++] = v;
    return false;
}

int* dfs_topological_with_cycle(Graph* g, bool* has_cycle) {
    Color* color = calloc(g->vertices, sizeof(Color));
    int* stack = malloc(g->vertices * sizeof(int));
    int stack_top = 0;

    *has_cycle = false;
    for (int i = 0; i < g->vertices; i++) {
        if (color[i] == WHITE) {
            if (dfs_topological_cycle_util(g, i, color, stack, &stack_top)) {
                *has_cycle = true;
                free(color);
                free(stack);
                return NULL;
            }
        }
    }

    int* result = malloc(g->vertices * sizeof(int));
    for (int i = 0; i < g->vertices; i++)
        result[i] = stack[g->vertices - 1 - i];

    free(color);
    free(stack);
    return result;
}

/* =============================================================================
 * 5. Find All Topological Sorts
 * ============================================================================= */

void all_topological_sorts_util(Graph* g, bool visited[], int in_degree[],
                                 int result[], int idx, int* count) {
    bool found = false;

    for (int v = 0; v < g->vertices; v++) {
        if (!visited[v] && in_degree[v] == 0) {
            visited[v] = true;
            result[idx] = v;

            /* Decrease in-degree of adjacent vertices */
            AdjNode* node = g->adj[v];
            while (node) {
                in_degree[node->vertex]--;
                node = node->next;
            }

            all_topological_sorts_util(g, visited, in_degree, result, idx + 1, count);

            /* Restore */
            visited[v] = false;
            node = g->adj[v];
            while (node) {
                in_degree[node->vertex]++;
                node = node->next;
            }

            found = true;
        }
    }

    if (!found && idx == g->vertices) {
        (*count)++;
        printf("      ");
        for (int i = 0; i < g->vertices; i++)
            printf("%d ", result[i]);
        printf("\n");
    }
}

int all_topological_sorts(Graph* g) {
    bool* visited = calloc(g->vertices, sizeof(bool));
    int* in_degree = calloc(g->vertices, sizeof(int));
    int* result = malloc(g->vertices * sizeof(int));
    int count = 0;

    /* Calculate in-degrees */
    for (int v = 0; v < g->vertices; v++) {
        AdjNode* node = g->adj[v];
        while (node) {
            in_degree[node->vertex]++;
            node = node->next;
        }
    }

    all_topological_sorts_util(g, visited, in_degree, result, 0, &count);

    free(visited);
    free(in_degree);
    free(result);
    return count;
}

/* =============================================================================
 * 6. Practical: Course Schedule
 * ============================================================================= */

bool can_finish_courses(int num_courses, int prerequisites[][2], int num_prereqs) {
    Graph* g = graph_create(num_courses);

    for (int i = 0; i < num_prereqs; i++) {
        graph_add_edge(g, prerequisites[i][1], prerequisites[i][0]);
    }

    bool has_cycle;
    int* result = kahn_topological_sort(g, &has_cycle);

    free(result);
    graph_free(g);

    return !has_cycle;
}

int* find_course_order(int num_courses, int prerequisites[][2], int num_prereqs,
                       int* order_size) {
    Graph* g = graph_create(num_courses);

    for (int i = 0; i < num_prereqs; i++) {
        graph_add_edge(g, prerequisites[i][1], prerequisites[i][0]);
    }

    bool has_cycle;
    int* result = kahn_topological_sort(g, &has_cycle);

    graph_free(g);

    if (has_cycle) {
        free(result);
        *order_size = 0;
        return NULL;
    }

    *order_size = num_courses;
    return result;
}

/* =============================================================================
 * 7. Practical: Task Scheduling
 * ============================================================================= */

int* task_scheduling(int num_tasks, int dependencies[][2], int num_deps,
                     int task_times[], int* completion_times) {
    Graph* g = graph_create(num_tasks);

    for (int i = 0; i < num_deps; i++) {
        graph_add_edge(g, dependencies[i][0], dependencies[i][1]);
    }

    bool has_cycle;
    int* order = kahn_topological_sort(g, &has_cycle);

    if (has_cycle) {
        graph_free(g);
        free(order);
        return NULL;
    }

    /* Calculate earliest completion times */
    for (int i = 0; i < num_tasks; i++)
        completion_times[i] = task_times[i];

    for (int i = 0; i < num_tasks; i++) {
        int v = order[i];
        AdjNode* node = g->adj[v];
        while (node) {
            int new_time = completion_times[v] + task_times[node->vertex];
            if (new_time > completion_times[node->vertex])
                completion_times[node->vertex] = new_time;
            node = node->next;
        }
    }

    graph_free(g);
    return order;
}

/* =============================================================================
 * Test
 * ============================================================================= */

void print_array(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
}

int main(void) {
    printf("============================================================\n");
    printf("Topological Sort Examples\n");
    printf("============================================================\n");

    /* 1. Kahn's Algorithm */
    printf("\n[1] Kahn's Algorithm (BFS)\n");
    Graph* g1 = graph_create(6);
    graph_add_edge(g1, 5, 2);
    graph_add_edge(g1, 5, 0);
    graph_add_edge(g1, 4, 0);
    graph_add_edge(g1, 4, 1);
    graph_add_edge(g1, 2, 3);
    graph_add_edge(g1, 3, 1);

    bool has_cycle;
    int* result = kahn_topological_sort(g1, &has_cycle);
    printf("    Graph: 5->2, 5->0, 4->0, 4->1, 2->3, 3->1\n");
    printf("    Topological sort: ");
    if (!has_cycle) print_array(result, 6);
    printf("\n");
    free(result);

    /* 2. DFS-based */
    printf("\n[2] DFS-based Topological Sort\n");
    result = dfs_topological_sort(g1);
    printf("    Topological sort: ");
    print_array(result, 6);
    printf("\n");
    free(result);

    /* 3. Cycle Detection */
    printf("\n[3] Cycle Detection\n");
    Graph* g2 = graph_create(3);
    graph_add_edge(g2, 0, 1);
    graph_add_edge(g2, 1, 2);
    graph_add_edge(g2, 2, 0);  /* Cycle */

    result = dfs_topological_with_cycle(g2, &has_cycle);
    printf("    Graph 0->1->2->0 (cycle): %s\n",
           has_cycle ? "cycle detected" : "DAG");
    if (result) free(result);
    graph_free(g2);

    /* 4. All Topological Sorts */
    printf("\n[4] All Topological Sorts\n");
    Graph* g3 = graph_create(4);
    graph_add_edge(g3, 0, 1);
    graph_add_edge(g3, 0, 2);
    graph_add_edge(g3, 1, 3);
    graph_add_edge(g3, 2, 3);

    printf("    Graph: 0->1, 0->2, 1->3, 2->3\n");
    printf("    All orderings:\n");
    int count = all_topological_sorts(g3);
    printf("    Total: %d\n", count);
    graph_free(g3);

    /* 5. Course Schedule */
    printf("\n[5] Course Schedule\n");
    int prereqs[][2] = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
    int order_size;
    int* course_order = find_course_order(4, prereqs, 4, &order_size);

    printf("    Courses: 0, 1, 2, 3\n");
    printf("    Prerequisites: 1<-0, 2<-0, 3<-1, 3<-2\n");
    if (course_order) {
        printf("    Course order: ");
        print_array(course_order, order_size);
        printf("\n");
        free(course_order);
    }

    /* 6. Task Scheduling */
    printf("\n[6] Task Scheduling\n");
    int deps[][2] = {{0, 1}, {0, 2}, {1, 3}, {2, 3}};
    int times[] = {2, 3, 4, 1};  /* Time required for each task */
    int completion[4];

    int* task_order = task_scheduling(4, deps, 4, times, completion);
    printf("    Task times: [2, 3, 4, 1]\n");
    printf("    Dependencies: 0->1, 0->2, 1->3, 2->3\n");
    if (task_order) {
        printf("    Completion times: ");
        for (int i = 0; i < 4; i++)
            printf("%d ", completion[i]);
        printf("\n");
        free(task_order);
    }

    graph_free(g1);

    /* 7. Algorithm Comparison */
    printf("\n[7] Algorithm Comparison\n");
    printf("    | Method    | Time       | Features                 |\n");
    printf("    |-----------|------------|--------------------------|\n");
    printf("    | Kahn(BFS) | O(V + E)   | Easy cycle detection     |\n");
    printf("    | DFS       | O(V + E)   | Simple implementation    |\n");
    printf("    | All sorts | O(V! * E)  | Enumerates all cases     |\n");

    printf("\n============================================================\n");

    return 0;
}
